pkg/ccl/storageccl/external_sst_reader.go

```diff
@@ -76,44 +90,73 @@ func ExternalSSTReader(
 				return nil, err
 			}
 		}
-		return storage.NewMemSSTIterator(content, false)
+		inMemorySSTs = append(inMemorySSTs, content)
 	}
+	return storage.NewPebbleMultiMemSSTIterator(inMemorySSTs, false, iterOps)
+}
 
-	raw := &sstReader{
-		ctx:  ctx,
-		sz:   sizeStat(sz),
-		body: f,
-		openAt: func(offset int64) (ioctx.ReadCloserCtx, error) {
-			reader, _, err := e.ReadFileAt(ctx, basename, offset)
-			return reader, err
-		},
+// ExternalSSTReader returns a PebbleSSTIterator for the SSTs in external storage,
+// optionally decrypting with the supplied parameters.
+//
+// ctx is captured and used throughout the life of the returned iterator, until
+// the iterator's Close() method is called.
+func ExternalSSTReader(
+	ctx context.Context,
+	e []cloud.ExternalStorage,
+	filenames []string,
+	encryption *roachpb.FileEncryptionOptions,
+	iterOps storage.IterOptions,
+) (storage.SimpleMVCCIterator, error) {
+	if len(e) != len(filenames) {
+		return nil, errors.New("must supply an external store for each file")
+	}
+	if !remoteSSTs.Get(&e[0].Settings().SV) {
+		return newMemPebbleSSTReader(ctx, e, filenames, encryption, iterOps)
 	}
+	remoteCacheSize := remoteSSTSuffixCacheSize.Get(&e[0].Settings().SV)
+	readers := make([]sstable.ReadableFile, 0, len(filenames))
 
-	var reader sstable.ReadableFile = raw
+	for i, basename := range filenames {
+		// prevent capturing the loop variables by reference when defining openAt below.
+		basename := basename
+		i := i
 
-	if encryption != nil {
-		r, err := decryptingReader(raw, encryption.Key)
+		f, sz, err := getFileWithRetry(ctx, basename, e[i])
 		if err != nil {
-			f.Close(ctx)
 			return nil, err
 		}
-		reader = r
-	} else {
-		// We only explicitly buffer the suffix of the file when not decrypting as
-		// the decrypting reader has its own internal block buffer.
-		if err := raw.readAndCacheSuffix(remoteSSTSuffixCacheSize.Get(&e.Settings().SV)); err != nil {
-			f.Close(ctx)
-			return nil, err
+
+		raw := &sstReader{
+			ctx:  ctx,
+			sz:   sizeStat(sz),
+			body: f,
+			openAt: func(offset int64) (ioctx.ReadCloserCtx, error) {
+				reader, _, err := e[i].ReadFileAt(ctx, basename, offset)
+				return reader, err
+			},
 		}
-	}
 
-	iter, err := storage.NewSSTIterator(reader)
-	if err != nil {
-		reader.Close()
-		return nil, err
-	}
+		var reader sstable.ReadableFile
 
-	return iter, nil
+		if encryption != nil {
+			r, err := decryptingReader(raw, encryption.Key)
+			if err != nil {
+				f.Close(ctx)
+				return nil, err
+			}
+			reader = r
+		} else {
+			// We only explicitly buffer the suffix of the file when not decrypting as
+			// the decrypting reader has its own internal block buffer.
+			if err := raw.readAndCacheSuffix(remoteCacheSize); err != nil {
+				f.Close(ctx)
+				return nil, err
+			}
+			reader = raw
+		}
+		readers = append(readers, reader)
+	}
+	return storage.NewPebbleSSTIterator(readers, iterOps)
```

Comment:
Note the order here in case there are conflicting keys. If the exact same Pebble key (that is, the same key/timestamp combination) exists in multiple files, the first one takes precedence. Not sure if we prevent that from happening, or what order the input files are in, but worth considering.

Reply:
Added a note to the function docstring to reflect this.